lib/Doctrine/DBAL/Driver/IBMDB2/DB2Statement.php

```diff
@@ -213,20 +215,28 @@ public function fetch($fetchMode = null)
                     $ctorArgs  = isset($args[2]) ? $args[2] : array();
                 }
 
-                $result = db2_fetch_object($this->_stmt);
+                $result = @db2_fetch_object($this->_stmt);
 
                 if ($result instanceof \stdClass) {
                     $result = $this->castObject($result, $className, $ctorArgs);
                 }
-
-                return $result;
+                break;
             case \PDO::FETCH_NUM:
-                return db2_fetch_array($this->_stmt);
+                $result = @db2_fetch_array($this->_stmt);
+                break;
             case \PDO::FETCH_OBJ:
-                return db2_fetch_object($this->_stmt);
+                $result = @db2_fetch_object($this->_stmt);
+                break;
             default:
                 throw new DB2Exception("Given Fetch-Style " . $fetchMode . " is not supported.");
         }
+
+        $error = db2_stmt_errormsg();
```

Comment:
What if I simply move the original contents of the `fetch` method to a private method `doFetch`, and make `fetch` call `doFetch` and check the result?

```php
    public function fetch($fetchMode = null)
    {
        $result = call_user_func_array(array($this, 'doFetch'), func_get_args());

        list($errorMessage, $errorCode) = $this->errorInfo();
        if (false === $result && !empty($errorMessage)) {
            throw [...]
        }

        return $result;
    }
```

That way, the original code (which wasn't unit tested anyway) is left alone, and the new, testable code wraps it.